src/Catalyst.Core.Modules.Ledger/Ledger.cs

```diff
@@ -0,0 +1,182 @@
+            _cryptoContext = new FfiWrapper();
+        }
+
+        private void FlushTransactionsFromDelta()
+        {
+            var transactionsToFlush = _mempool.Repository.GetAll(); //TOD0 no get alls
+            _mempool.Repository.Delete(transactionsToFlush);
+        }
+
+        /// <inheritdoc />
+        public bool SaveAccountState(Account account)
+        {
+            Guard.Argument(account, nameof(account)).NotNull();
+
+            try
+            {
+                Accounts.Add(account);
+                return true;
+            }
+            catch (Exception e)
+            {
+                _logger.Error(e, "Failed to add account state to the Ledger");
+                return false;
+            }
+        }
+
+        /// <inheritdoc />
+        public void Update(Cid deltaHash)
+        {
+            try
+            {
+                lock (_synchronisationLock)
+                {
+                    var chainedDeltaHashes = _synchroniser
+                       .CacheDeltasBetween(LatestKnownDelta, deltaHash, CancellationToken.None)
+                       .Reverse()
+                       .ToList();
+
+                    if (!Equals(chainedDeltaHashes.First(), LatestKnownDelta))
+                    {
+                        _logger.Warning(
+                            "Failed to walk back the delta chain to {LatestKnownDelta}, giving up ledger update.",
+                            LatestKnownDelta);
+                        return;
+                    }
+
+                    foreach (var chainedDeltaHash in chainedDeltaHashes)
+                    {
+                        UpdateLedgerFromDelta(chainedDeltaHash);
+                    }
+                }
+
+                //https://github.com/catalyst-network/Catalyst.Node/issues/871
+                FlushTransactionsFromDelta();
+            }
+            catch (Exception exception)
+            {
+                _logger.Error(exception, "Failed to update the ledger using the delta with hash {deltaHash}",
+                    deltaHash);
+            }
+        }
+
+        private void UpdateLedgerFromDelta(Cid deltaHash)
+        {
+            if (!_synchroniser.DeltaCache.TryGetOrAddConfirmedDelta(deltaHash, out var nextDeltaInChain))
+            {
+                _logger.Warning(
+                    "Failed to retrieve Delta with hash {hash} from the Dfs, ledger has not been updated.", deltaHash);
+                return;
+            }
+
+            foreach (var entry in nextDeltaInChain.PublicEntries)
+            {
+                UpdateLedgerAccountFromEntry(entry);
+            }
+
+            LatestKnownDelta = deltaHash;
+        }
+
+        private void UpdateLedgerAccountFromEntry(PublicEntry entry)
+        {
+            var pubKey = _cryptoContext.GetPublicKeyFromBytes(entry.Base.ReceiverPublicKey.ToByteArray());
+
+            //todo: get an address from the key using the Account class from Common lib
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/app/catalyst-network/Catalyst.Node/pullRequest?prid=4318201)